code/modules/integrated_electronics/core/helpers.dm

```diff
@@ -21,6 +21,9 @@
 
 
 /obj/item/integrated_circuit/proc/set_pin_data(pin_type, pin_number, datum/new_data)
+	if(islist(new_data))
+		for(var/i in 1 to length(new_data))
```

Comment:
So we need to iterate across all the list just in case we'll meet some obj refs.
List processing in circuits already slow enought.

Reply:
It would have to do this somewhere regardless, whenever atoms get added to a list. How else would you suggest it be handled?

Reply:
handle it in circuits. It means more bulky code of circuits itself, but better performance.

Reply:
It sounds like a lot of unnecessary duplicate code. Is the performance cost of going through a list to see if an item can be WEAKREFed or not really so severe? Would it be better if I used for(var/i in new_data) here instead?

Reply:
It is really severe.Since circuit is copying list every fucking time it pushes list,You want to add one more cycle to every this operation, including cases, where you shouldn't deal with refs at all.Don't hate srvr so much. you can make special proc for this fuckery to reduce amount of dupe code.But use it only where you can occur refs, needed to be weakrefed.

Reply:
Well, okay. I think, I know solution. will implement myself later. This is fine for now